src/Raven.Server/Documents/ETL/Relational/RelationalWriters/RelationalWriterBase.cs

```diff
@@ -0,0 +1,748 @@
+
+                    var tableMetrics = _sqlMetrics.GetTableMetrics(tableName);
+                    tableMetrics.DeleteActionsMeter.MarkSingleThreaded(1);
+
+                    if (elapsedMilliseconds > LongStatementWarnThresholdInMs)
+                    {
+                        HandleSlowSql(elapsedMilliseconds, stmt);
+                    }
+                }
+            }
+        }
+
+        return deleted;
+    }
+
+    private void HandleSlowSql(long elapsedMilliseconds, string stmt)
+    {
+        if (Logger.IsInfoEnabled)
+            Logger.Info($"[{_etlName}] Slow SQL detected. Execution took: {elapsedMilliseconds:#,#;;0}ms, statement: {stmt}");
+
+        _statistics.RecordSlowSql(new SlowSqlStatementInfo { Date = SystemTime.UtcNow, Duration = elapsedMilliseconds, Statement = stmt });
+    }
+
+    protected abstract bool ShouldQuoteTables();
+
+    private string GetTableNameString(string tableName)
+    {
+        if (ShouldQuoteTables())
+        {
+            return string.Join(".", tableName.Split('.').Select(_commandBuilder.QuoteIdentifier).ToArray());
+        }
+
+        return tableName;
+    }
+
+    public static string SanitizeSqlValue(string sqlValue)
+    {
+        return sqlValue.Replace("'", "''");
+    }
+    
+    protected abstract string GetParameterNameForDbParameter(string paramName);
+
+    protected abstract string GetParameterNameForCommandString(string targetParamName, bool parseJson);
+    
+    protected abstract void EnsureParamTypeSupportedByDbProvider(DbParameter parameter);
+
+    protected abstract void SetPrimaryKeyParamValue(ToRelationalItem itemToReplicate, DbParameter pkParam);
+
+    protected abstract string GetPostInsertIntoStartSyntax(ToRelationalItem itemToReplicate);
+    protected abstract string GetPostInsertIntoEndSyntax(ToRelationalItem itemToReplicate);
+    
+    protected abstract string GetPostDeleteSyntax(ToRelationalItem itemToDelete);
+
+    public RelationalWriteStats Write(RelationalTableWithRecords table, List<DbCommand> commands, CancellationToken token)
+    {
+        var stats = new RelationalWriteStats();
+        
+        var collectCommands = commands != null ? commands.Add : (System.Action<DbCommand>)null;
+
+        if (table.InsertOnlyMode == false && table.Deletes.Count > 0)
+        {
+            // first, delete all the rows that might already exist there
+            stats.DeletedRecordsCount = DeleteItems(table.TableName, table.DocumentIdColumn, ParametrizeDeletes, table.Deletes, collectCommands,
+                token);
+        }
+
+        if (table.Inserts.Count > 0)
+        {
+            stats.InsertedRecordsCount = InsertItems(table.TableName, table.DocumentIdColumn, table.Inserts, collectCommands, token);
+        }
+
+        return stats;
+    }
+
+    private static void SetProviderSpecificDbType(string dbTypeString, ref DbParameter colParam, SqlProvider? providerType)
```

Comment:
Should it be in base class?

Reply:
moved to `SqlDatabaseWriter.cs` as a internal method